File: app/src/main/java/br/com/menuvem/lojista/presentation/shopping/ShoppingListScreen.kt
package br.com.menuvem.lojista.presentation.shopping

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.outlined.ShoppingCart
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import br.com.menuvem.lojista.domain.model.StatusLista
import br.com.menuvem.lojista.presentation.components.EmptyState
import br.com.menuvem.lojista.presentation.components.ItemListaCard
import br.com.menuvem.lojista.presentation.components.formatarMoeda
import br.com.menuvem.lojista.ui.theme.PurplePrimary
import br.com.menuvem.lojista.ui.theme.SuccessGreen
import br.com.menuvem.lojista.ui.theme.YellowSecondary
import java.time.format.DateTimeFormatter

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ShoppingListScreen(
    listaId: Long,
    onBack: () -> Unit,
    viewModel: ShoppingListViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    val snackbarHostState = remember { SnackbarHostState() }

    LaunchedEffect(uiState.finishedSuccess) {
        if (uiState.finishedSuccess) {
            snackbarHostState.showSnackbar("Compra finalizada! Custos atualizados.")
            viewModel.clearSuccess()
        }
    }

    LaunchedEffect(uiState.error) {
        uiState.error?.let {
            snackbarHostState.showSnackbar(it)
        }
    }

    val lista = uiState.lista
    val isFinalized = lista?.status == StatusLista.FINALIZADA

    Scaffold(
        snackbarHost = { SnackbarHost(snackbarHostState) },
        topBar = {
            Column {
                // Cabeçalho com gradiente
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(
                            Brush.verticalGradient(
                                colors = listOf(PurplePrimary, PurplePrimary.copy(alpha = 0.85f))
                            )
                        )
                        .statusBarsPadding()
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            IconButton(onClick = onBack) {
                                Icon(
                                    Icons.Default.ArrowBack,
                                    contentDescription = "Voltar",
                                    tint = Color.White
                                )
                            }
                            Column(modifier = Modifier.weight(1f)) {
                                Text(
                                    text = lista?.nome ?: "Lista de Compras",
                                    style = MaterialTheme.typography.titleLarge,
                                    color = Color.White,
                                    fontWeight = FontWeight.Bold,
                                    maxLines = 1,
                                    overflow = TextOverflow.Ellipsis
                                )
                                lista?.dataCriacao?.let { data ->
                                    Text(
                                        text = data.format(DateTimeFormatter.ofPattern("dd/MM/yyyy")),
                                        style = MaterialTheme.typography.bodySmall,
                                        color = Color.White.copy(alpha = 0.75f)
                                    )
                                }
                            }
                            if (isFinalized) {
                                Surface(
                                    shape = MaterialTheme.shapes.small,
                                    color = SuccessGreen.copy(alpha = 0.25f)
                                ) {
                                    Row(
                                        modifier = Modifier.padding(horizontal = 10.dp, vertical = 4.dp),
                                        verticalAlignment = Alignment.CenterVertically,
                                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                                    ) {
                                        Icon(
                                            Icons.Default.CheckCircle,
                                            contentDescription = null,
                                            tint = Color.White,
                                            modifier = Modifier.size(14.dp)
                                        )
                                        Text(
                                            text = "Finalizada",
                                            style = MaterialTheme.typography.labelSmall,
                                            color = Color.White
                                        )
                                    }
                                }
                            }
                        }

                        Spacer(modifier = Modifier.height(12.dp))

                        // Barra de progresso e totais
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween,
                            verticalAlignment = Alignment.Bottom
                        ) {
                            Column {
                                Text(
                                    text = "${uiState.itensComprados} de ${uiState.totalItens} itens",
                                    style = MaterialTheme.typography.bodySmall,
                                    color = Color.White.copy(alpha = 0.8f)
                                )
                                Text(
                                    text = "R$ ${formatarMoeda(uiState.totalGasto)}",
                                    style = MaterialTheme.typography.headlineSmall,
                                    color = Color.White,
                                    fontWeight = FontWeight.Bold
                                )
                            }
                            if (uiState.totalItens > 0) {
                                val progress = uiState.itensComprados.toFloat() / uiState.totalItens
                                CircularProgressIndicator(
                                    progress = { progress },
                                    color = Color.White,
                                    trackColor = Color.White.copy(alpha = 0.3f),
                                    modifier = Modifier.size(48.dp),
                                    strokeWidth = 4.dp
                                )
                            }
                        }

                        if (uiState.totalItens > 0) {
                            Spacer(modifier = Modifier.height(8.dp))
                            LinearProgressIndicator(
                                progress = {
                                    if (uiState.totalItens > 0)
                                        uiState.itensComprados.toFloat() / uiState.totalItens
                                    else 0f
                                },
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .height(4.dp),
                                color = Color.White,
                                trackColor = Color.White.copy(alpha = 0.3f)
                            )
                        }
                    }
                }
            }
        },
        floatingActionButton = {
            if (!isFinalized) {
                Column(
                    horizontalAlignment = Alignment.End,
                    verticalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    // FAB Finalizar (apenas se há itens comprados)
                    if (uiState.itensComprados > 0) {
                        FloatingActionButton(
                            onClick = viewModel::onShowFinishDialog,
                            containerColor = SuccessGreen,
                            contentColor = Color.White,
                            modifier = Modifier.size(48.dp)
                        ) {
                            Icon(Icons.Default.Done, contentDescription = "Finalizar compra")
                        }
                    }
                    // FAB Adicionar
                    FloatingActionButton(
                        onClick = viewModel::onShowAddSheet,
                        containerColor = YellowSecondary,
                        contentColor = Color.White
                    ) {
                        Icon(Icons.Default.Add, contentDescription = "Adicionar item")
                    }
                }
            }
        }
    ) { paddingValues ->
        if (uiState.isLoading) {
            Box(
                modifier = Modifier.fillMaxSize().padding(paddingValues),
                contentAlignment = Alignment.Center
            ) { CircularProgressIndicator(color = PurplePrimary) }
        } else if (uiState.itens.isEmpty()) {
            Box(
                modifier = Modifier.fillMaxSize().padding(paddingValues),
                contentAlignment = Alignment.Center
            ) {
                EmptyState(
                    icon = Icons.Outlined.ShoppingCart,
                    title = "Lista vazia",
                    subtitle = "Toque em + para adicionar os primeiros itens"
                )
            }
        } else {
            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                contentPadding = PaddingValues(
                    start = 12.dp, end = 12.dp,
                    top = 8.dp, bottom = 100.dp
                ),
                verticalArrangement = Arrangement.spacedBy(6.dp)
            ) {
                // Itens pendentes
                val pendentes = uiState.itens.filter { !it.comprado }
                val comprados = uiState.itens.filter { it.comprado }

                if (pendentes.isNotEmpty()) {
                    item {
                        Text(
                            text = "A comprar (${pendentes.size})",
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.padding(vertical = 4.dp, horizontal = 4.dp)
                        )
                    }
                    items(pendentes, key = { it.id }) { item ->
                        ItemListaCard(
                            item = item,
                            tendencia = item.insumoId?.let { uiState.tendencias[it] },
                            onToggle = { checked, preco -> viewModel.toggleItem(item.id, checked, preco) },
                            onDelete = { viewModel.deleteItem(item) }
                        )
                    }
                }

                if (comprados.isNotEmpty()) {
                    item {
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(
                            text = "Comprados (${comprados.size})",
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.padding(vertical = 4.dp, horizontal = 4.dp)
                        )
                    }
                    items(comprados, key = { it.id }) { item ->
                        ItemListaCard(
                            item = item,
                            tendencia = item.insumoId?.let { uiState.tendencias[it] },
                            onToggle = { checked, preco -> viewModel.toggleItem(item.id, checked, preco) },
                            onDelete = { viewModel.deleteItem(item) }
                        )
                    }
                }
            }
        }

        // Bottom Sheet
        if (uiState.showAddSheet) {
            AddItemBottomSheet(
                searchQuery = uiState.searchQuery,
                searchResults = uiState.searchResults,
                isSearching = uiState.isSearching,
                onSearchChanged = viewModel::onSearchQueryChanged,
                onAddFreeItem = { nome, qtd, unidade, preco ->
                    viewModel.addItem(nome, qtd, unidade, preco)
                },
                onAddInsumo = { insumo, qtd ->
                    viewModel.addItem(
                        nomeItem = insumo.nome,
                        quantidade = qtd,
                        unidade = insumo.unidadeCompra,
                        precoUnitario = insumo.custoAtual,
                        insumo = insumo
                    )
                },
                onDismiss = viewModel::onHideAddSheet
            )
        }

        // Dialog de confirmação
        if (uiState.showFinishDialog) {
            AlertDialog(
                onDismissRequest = viewModel::onDismissFinishDialog,
                icon = { Icon(Icons.Default.ShoppingBag, contentDescription = null) },
                title = { Text("Finalizar compra?") },
                text = {
                    Text(
                        "Os preços dos ${uiState.itensComprados} itens comprados serão " +
                        "registrados e os insumos vinculados terão seus custos atualizados automaticamente."
                    )
                },
                confirmButton = {
                    Button(
                        onClick = viewModel::finalizarCompra,
                        colors = ButtonDefaults.buttonColors(containerColor = SuccessGreen)
                    ) {
                        if (uiState.isFinishing) {
                            CircularProgressIndicator(
                                color = Color.White,
                                modifier = Modifier.size(20.dp),
                                strokeWidth = 2.dp
                            )
                        } else {
                            Text("Finalizar")
                        }
                    }
                },
                dismissButton = {
                    TextButton(onClick = viewModel::onDismissFinishDialog) {
                        Text("Cancelar")
                    }
                }
            )
        }
    }
}

File: lib/presentation/shopping/shopping_list_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:intl/intl.dart';

import '../../domain/model/lista_compras.dart';
import '../../theme/app_theme.dart';
import '../components/back_or_home_button.dart';
import '../components/empty_state.dart';
import '../components/formatters.dart';
import '../components/responsive.dart';
import 'add_item_bottom_sheet.dart';
import 'item_lista_card.dart';
import 'shopping_list_controller.dart';

class ShoppingListScreen extends ConsumerWidget {
  const ShoppingListScreen({super.key, required this.listaId});

  final int listaId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final dataAsync = ref.watch(shoppingListDataProvider(listaId));
    final actions = ref.read(shoppingListActionsProvider(listaId));

    return Scaffold(
      appBar: AppBar(
        leading: const BackOrHomeButton(),
        title: dataAsync.maybeWhen(
          data: (d) => Text(d.lista?.nome ?? 'Lista de Compras'),
          orElse: () => const Text('Lista de Compras'),
        ),
      ),
      floatingActionButton: dataAsync.maybeWhen(
        data: (data) {
          if (data.lista?.status == StatusLista.finalizada) return null;
          return Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              if (data.itensComprados > 0)
                Padding(
                  padding: const EdgeInsets.only(bottom: 12),
                  child: FloatingActionButton(
                    heroTag: 'finish',
                    backgroundColor: successGreen,
                    foregroundColor: Colors.white,
                    onPressed: () => _showFinishDialog(
                      context,
                      ref,
                      actions,
                      data.itensComprados,
                    ),
                    child: const Icon(Icons.done),
                  ),
                ),
              FloatingActionButton(
                heroTag: 'add',
                backgroundColor: yellowSecondary,
                foregroundColor: Colors.white,
                onPressed: () => showResponsiveFormSheet<void>(
                  context,
                  builder: (_) => AddItemBottomSheet(
                    onAdd:
                        ({
                          required nomeItem,
                          required quantidade,
                          required unidade,
                          precoUnitario = 0.0,
                          insumo,
                        }) => actions.addItem(
                          nomeItem: nomeItem,
                          quantidade: quantidade,
                          unidade: unidade,
                          precoUnitario: precoUnitario,
                          insumo: insumo,
                        ),
                  ),
                ),
                child: const Icon(Icons.add),
              ),
            ],
          );
        },
        orElse: () => null,
      ),
      body: MaxWidthCenter(
        child: dataAsync.when(
          loading: () => const Center(
            child: CircularProgressIndicator(color: purplePrimary),
          ),
          error: (e, _) => Center(child: Text('Erro: $e')),
          data: (data) {
            if (data.itens.isEmpty) {
              return const EmptyState(
                icon: Icons.shopping_cart_outlined,
                title: 'Lista vazia',
                subtitle: 'Toque em + para adicionar os primeiros itens',
              );
            }
            final pendentes = data.itens.where((i) => !i.comprado).toList();
            final comprados = data.itens.where((i) => i.comprado).toList();
            return Column(
              children: [
                _HeaderResumo(
                  lista: data.lista,
                  comprados: data.itensComprados,
                  total: data.totalItens,
                  totalGasto: data.totalGasto,
                ),
                Expanded(
                  child: ListView(
                    padding: const EdgeInsets.fromLTRB(12, 8, 12, 100),
                    children: [
                      if (pendentes.isNotEmpty) ...[
                        Padding(
                          padding: const EdgeInsets.symmetric(
                            vertical: 4,
                            horizontal: 4,
                          ),
                          child: Text(
                            'A comprar (${pendentes.length})',
                            style: Theme.of(context).textTheme.labelMedium,
                          ),
                        ),
                        ...pendentes.map(
                          (item) => Padding(
                            padding: const EdgeInsets.only(bottom: 6),
                            child: ItemListaCard(
                              item: item,
                              tendencia: item.insumoId != null
                                  ? data.tendencias[item.insumoId]
                                  : null,
                              onToggle: (checked, preco) =>
                                  actions.toggleItem(item.id, checked, preco),
                              onDelete: () => actions.deleteItem(item),
                            ),
                          ),
                        ),
                      ],
                      if (comprados.isNotEmpty) ...[
                        Padding(
                          padding: const EdgeInsets.symmetric(
                            vertical: 4,
                            horizontal: 4,
                          ),
                          child: Text(
                            'Comprados (${comprados.length})',
                            style: Theme.of(context).textTheme.labelMedium,
                          ),
                        ),
                        ...comprados.map(
                          (item) => Padding(
                            padding: const EdgeInsets.only(bottom: 6),
                            child: ItemListaCard(
                              item: item,
                              tendencia: item.insumoId != null
                                  ? data.tendencias[item.insumoId]
                                  : null,
                              onToggle: (checked, preco) =>
                                  actions.toggleItem(item.id, checked, preco),
                              onDelete: () => actions.deleteItem(item),
                            ),
                          ),
                        ),
                      ],
                    ],
                  ),
                ),
              ],
            );
          },
        ),
      ),
    );
  }

  void _showFinishDialog(
    BuildContext context,
    WidgetRef ref,
    ShoppingListActions actions,
    int itensComprados,
  ) {
    showDialog<void>(
      context: context,
      builder: (context) =>
          _FinishDialog(itensComprados: itensComprados, actions: actions),
    );
  }
}

class _FinishDialog extends StatefulWidget {
  const _FinishDialog({required this.itensComprados, required this.actions});
  final int itensComprados;
  final ShoppingListActions actions;

  @override
  State<_FinishDialog> createState() => _FinishDialogState();
}

class _FinishDialogState extends State<_FinishDialog> {
  bool _isFinishing = false;

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      icon: const Icon(Icons.shopping_bag),
      title: const Text('Finalizar compra?'),
      content: Text(
        'Os preços dos ${widget.itensComprados} itens comprados serão registrados e os '
        'insumos vinculados terão seus custos atualizados automaticamente.',
      ),
      actions: [
        TextButton(
          onPressed: () => Navigator.of(context).pop(),
          child: const Text('Cancelar'),
        ),
        FilledButton(
          style: FilledButton.styleFrom(backgroundColor: successGreen),
          onPressed: _isFinishing
              ? null
              : () async {
                  setState(() => _isFinishing = true);
                  await widget.actions.finalizarCompra();
                  if (context.mounted) {
                    Navigator.of(context).pop();
                    ScaffoldMessenger.of(context).showSnackBar(
                      const SnackBar(
                        content: Text('Compra finalizada! Custos atualizados.'),
                      ),
                    );
                  }
                },
          child: _isFinishing
              ? const SizedBox(
                  width: 20,
                  height: 20,
                  child: CircularProgressIndicator(
                    strokeWidth: 2,
                    color: Colors.white,
                  ),
                )
              : const Text('Finalizar'),
        ),
      ],
    );
  }
}

class _HeaderResumo extends StatelessWidget {
  const _HeaderResumo({
    required this.lista,
    required this.comprados,
    required this.total,
    required this.totalGasto,
  });
  final ListaCompras? lista;
  final int comprados;
  final int total;
  final double totalGasto;

  @override
  Widget build(BuildContext context) {
    final progress = total > 0 ? comprados / total : 0.0;
    final formatter = DateFormat('dd/MM/yyyy');
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        gradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          colors: [purplePrimary, purplePrimary.withValues(alpha: 0.85)],
        ),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          if (lista != null)
            Text(
              formatter.format(lista!.dataCriacao),
              style: Theme.of(context).textTheme.bodySmall?.copyWith(
                color: Colors.white.withValues(alpha: 0.75),
              ),
            ),
          if (lista?.status == StatusLista.finalizada)
            Container(
              margin: const EdgeInsets.only(top: 4),
              padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
              decoration: BoxDecoration(
                color: successGreen.withValues(alpha: 0.25),
                borderRadius: BorderRadius.circular(8),
              ),
              child: const Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Icon(Icons.check_circle, size: 14, color: Colors.white),
                  SizedBox(width: 4),
                  Text(
                    'Finalizada',
                    style: TextStyle(color: Colors.white, fontSize: 12),
                  ),
                ],
              ),
            ),
          const SizedBox(height: 8),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    '$comprados de $total itens',
                    style: Theme.of(context).textTheme.bodySmall?.copyWith(
                      color: Colors.white.withValues(alpha: 0.8),
                    ),
                  ),
                  Text(
                    'R\$ ${formatarMoeda(totalGasto)}',
                    style: Theme.of(context).textTheme.headlineSmall?.copyWith(
                      color: Colors.white,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                ],
              ),
              if (total > 0)
                SizedBox(
                  width: 44,
                  height: 44,
                  child: CircularProgressIndicator(
                    value: progress,
                    color: Colors.white,
                    backgroundColor: Colors.white.withValues(alpha: 0.3),
                    strokeWidth: 4,
                  ),
                ),
            ],
          ),
          if (total > 0) ...[
            const SizedBox(height: 8),
            ClipRRect(
              borderRadius: BorderRadius.circular(4),
              child: LinearProgressIndicator(
                value: progress,
                minHeight: 4,
                color: Colors.white,
                backgroundColor: Colors.white.withValues(alpha: 0.3),
              ),
            ),
          ],
        ],
      ),
    );
  }
}
